src/ParserRuleContext.ts

```diff
@@ -46,7 +46,26 @@ export class ParserRuleContext extends RuleContext {
 	 *  operation because we don't the need to track the details about
 	 *  how we parse this rule.
 	 */
-	children?: ParseTree[];
+
+
+	_children?: Array<ParseTree>;
+	static _emptyChildren = [];
+
+	/** Access children for read
+	 * Avoids creation of extra empty arrays by using a shared _emptyChildren array
+	 */
+	get children() : ReadonlyArray<ParseTree> {
+		let result = this._children;
+		return result ? result : ParserRuleContext._emptyChildren;
+	}
+
+	/** Access the children in a read/write way.
+	 * This will create the _children array if needed.
+	 */
+	get mutableChildren() {
+		if (!this._children) this._children = [];
+		return this._children as Array<ParseTree>;
+	}
```

Comment:
Hmm, this is the same approach as cocoa is using (NSArray, NSMutableArray etc.), but the question here is: what do you want to make the default? If you use immutable objects as default (by returning that on a call to children) you should be consequent and do that everywhere in the runtime. Not sure if you really want that. Or you can make the `children` property to return a mutable array and have another property `immutableChildren`. However, wouldn't it be then already enough to assign the result of `children` to a const var to make it immutable? To be honest I fail to see the benefit of having a property returning a read-only list.